feat(core): add API to provide CSP nonce for inline stylesheets

crisbeto opened the pull request:
Angular uses inline styles to insert the styles associated with a component. This violates the strict styles [Content Security Policy](https://web.dev/strict-csp/) which doesn't allow inline styles by default. One way to allow the styles to be applied is to set a `nonce` attribute on them, but because the code for inserting the stylesheets is deep inside the framework, users weren't able to provide it without accessing private APIs.

These changes add a new `CSP_NONCE` injection token that will allow users to provide a nonce, if their app is using CSP. If the token isn't provided, the framework will look for an `ngCspNonce` attribute on the app's root node instead. The latter approach is provided as a convenience for apps that render the `index.html` through a server, e.g. `<app ngCspNonce="{% randomNonceAddedByTheServer %}"></app>`.

This PR addresses adding the nonce to framework-generated styles. There will be follow-up PRs that add support for it in critical CSS tags in the CLI, and in Angular Material.

Fixes #6361.

crisbeto commented on packages/platform-server/src/styles_host.ts:
```diff
@@ -7,26 +7,34 @@
   constructor(@Inject(DOCUMENT) doc: any, @Inject(APP_ID) private appId: string) {
     super();
-    this.head = doc.getElementsByTagName('head')[0];
+    this.head = doc.head || doc.getElementsByTagName('head')[0];
```
Unrelated change, I just saw that we could get the `head` directly without querying the DOM. I left the old approach as a fallback in case it was there as a workaround for Domino.

crisbeto commented on packages/platform-browser/src/dom/dom_renderer.ts:
```diff
@@ -91,7 +91,8 @@ export class DomRendererFactory2 implements RendererFactory2, OnDestroy {
       @Inject(APP_ID) private appId: string,
-      @Inject(REMOVE_STYLES_ON_COMPONENT_DESTROY) private removeStylesOnCompDestory: boolean) {
+      @Inject(REMOVE_STYLES_ON_COMPONENT_DESTROY) private removeStylesOnCompDestory: boolean,
+      @Inject(CSP_NONCE) @Optional() private nonce?: string|null) {
```
This uses `@Inject` instead of `inject`, because we had a bunch of tests that were constructing classes manually which lacked a DI context.

jessicajaniuk reviewed (approved):
Aside from @AndrewKushnir's comment, this looks good to me

reviewed-for: fw-core, fw-platform-server, public-api

dylhunn reviewed (approved):
reviewed-for: public-api

alxhub reviewed (approved):
Reviewed-for: public-api

alan-agius4 commented on packages/platform-browser/src/dom/shared_styles_host.ts:
```diff
@@ -149,6 +150,11 @@ export class SharedStylesHost implements OnDestroy {
   private addStyleToHost(host: Node, style: string): void {
     const styleEl = this.getStyleElement(host, style);
 
+    if (this.nonce) {
```
Would we great if you can move this https://github.com/angular/angular/blob/7831a7a95f48c3942b7b63307c1df683c9c52f3d/packages/platform-browser/src/dom/shared_styles_host.ts#L144 where all the style tag manipulation happen.

alan-agius4 commented on packages/platform-browser/src/dom/shared_styles_host.ts:
```diff
@@ -149,6 +150,11 @@ export class SharedStylesHost implements OnDestroy {
     const styleEl = this.getStyleElement(host, style);
 
+    if (this.nonce) {
+      // Uses a keyed write to avoid issues with property minification.
```
OOC, does closure minify this property?

crisbeto replied:
I'm not sure, but I wanted to keep it safe since I've seen it minify things like `el.style.foo` before.

crisbeto committed:
feat(core): add API to provide CSP nonce for inline stylesheets

Angular uses inline styles to insert the styles associated with a component. This violates the strict styles [Content Security Policy](https://web.dev/strict-csp/) which doesn't allow inline styles by default. One way to allow the styles to be applied is to set a `nonce` attribute on them, but because the code for inserting the stylesheets is deep inside the framework, users weren't able to provide it without accessing private APIs.

These changes add a new `CSP_NONCE` injection token that will allow users to provide a nonce, if their app is using CSP. If the token isn't provided, the framework will look for an `ngCspNonce` attribute on the app's root node instead. The latter approach is provided as a convenience for apps that render the `index.html` through a server, e.g. `<app ngCspNonce="{% randomNonceAddedByTheServer %}"></app>`.

This PR addresses adding the nonce to framework-generated styles. There will be follow-up PRs that add support for it in critical CSS tags in the CLI, and in Angular Material.

Fixes #6361.

pkozlowski-opensource commented:
This PR was merged into the repository by commit 17e9862653758ebdbd29771cd4ec8a59436497d6.
